Config Validation Unit tests

arthur740212 opened the pull request:
Provides attributes to mark test types on a config field.
Tests for fields in `EOSConfig`
Generic FieldValidator that grabs the marked test types of a field

WispyMouse committed:
tests(config): Scaffolding for baseline unit test?

WispyMouse committed:
feat: EOSConfig Validation Attributes

WispyMouse committed:
feat: EOS Config test scaffolding

WispyMouse committed:
tests: Check that expected successes work

WispyMouse committed:
tests: All tests passing

arthur740212 committed:
feat(config,test) : generalize field validator to be used beyond EOSConfig

WispyMouse commented on Assets/Tests/PlayMode/EOSConfigTests.cs:
```diff
@@ -218,7 +218,7 @@ public void DeploymentOverride_MustBeValid()
             };
 
-            if (!config.TryGetFailingValidatorAttributes(out List<FieldValidatorFailure> failingAttributes))
+            if (!FieldValidator.TryGetFailingValidatorAttributes(config.sandboxDeploymentOverrides[0], out List<FieldValidatorFailure> failingAttributes))
```
Here, I think we should run this on both the config and on the targeted override. Running it on the config ensures that the config is checking its innards for more problems.

arthur740212 replied:
This will be updated alongside the recursive validate task

WispyMouse commented on com.playeveryware.eos/Runtime/Core/Config/Attributes/FieldValidator.cs:
```diff
@@ -0,0 +1,64 @@
+            FailingAttribute = failingAttribute;
+            FailingMessage = failingMessage;
+        }
+    }
```
nit: add new line

WispyMouse commented on com.playeveryware.eos/Runtime/Core/Config/Attributes/FieldValidator.cs:
```diff
@@ -0,0 +1,64 @@
+    }
+    public class FieldValidator
+    {
+        public static bool TryGetFailingValidatorAttributes(object target, out List<FieldValidatorFailure> failingValidatorAttributes)
```
Something possibly lost from this change is that it isn't checking fields inside itself that also have attributes. Should the config validator be able to do that all in one call? Is there a way we can recursively ask children for their attributes?

arthur740212 replied:
Yes, this is next task on deck.

paulhazen commented on com.playeveryware.eos/Runtime/Core/Config/Config.cs:
```diff
@@ -220,7 +220,7 @@ public static T Get<T>() where T : Config
         /// <returns>
```
Instead of being made `public`, if these things need to be utilized outside of the `Config` class, I believe that they should be moved to a shared location outside of where either is used. Perhaps they can be moved into the `EnumUtility` class itself?

arthur740212 replied:
It is reverted instead because it isn't used anywhere in the plugin

paulhazen reviewed (changes requested):
I started a review, but then noticed that there were quite a lot of differences between the branch being merged from and the `development` branch (about 125 commits).

I first noticed this looking at the layout of the `eos_plugin_for_unity/Assets/Tests/PlayMode/` directory, which in `development` looks like this:

![image](https://github.com/user-attachments/assets/fa65fa1a-60ff-420c-b17f-4b5c898874de)

I would imagine that the `EOSConfigTests` file that is added would be put under the `Config` directory.

Once I noticed that the branch was so far out of date, I stopped my review because I imagine a lot might be changed, and my feedback would be more clear afterwards.

arthur740212 committed:
fix(test,config) : fix compile errors following config field changes

arthur740212 committed:
fix(config,test) : Recursively validate members of a class

arthur740212 committed:
feat(config,test) : Expand attribute to validate members that are classes

paulhazen commented on com.playeveryware.eos/Runtime/Core/Config/Attributes/ExpandFieldAttribute.cs:
```diff
@@ -0,0 +1,35 @@
+    using System;
+
+    /// <summary>
+    /// This attribute is used to decorate a field member within a config class
+    /// that represents a path to a file.
```
This comment appears to be copied from another attribute. It should be updated to describe the purpose of this attribute - which isn't clear to me.

paulhazen commented on com.playeveryware.eos/Runtime/Core/Config/Attributes/GUIDFieldValidatorAttribute.cs:
```diff
@@ -0,0 +1,58 @@
+
+        public override bool FieldValueIsValid(object toValidate, out string configurationProblemMessage)
+        {
+            if (!(toValidate is string stringValue))
```
I _think_ you can actually change this to be:
```c#
if (toValidate is not string stringValue)
```

arthur740212 replied:
Yes, but it isn't necessary

paulhazen commented on com.playeveryware.eos/Runtime/Core/Config/Attributes/ParsesToAuthScopeFieldValidatorAttribute.cs:
```diff
@@ -0,0 +1,48 @@
+
+    public class ParsesToAuthScopeFieldValidatorAttribute : ParsesToEnumFieldValidatorAttribute
+    {
+        protected override bool CanParse(List<string> flags)
+        {
+            if (!AuthScopeFlagsExtensions.TryParse(flags, out AuthScopeFlags results))
+            {
+                return false;
+            }
+
+            return true;
+        }
```
This should no longer be needed - since the `List<string>` type of Enums disappeared for `EOSConfig`.

arthur740212 replied:
This can be removed, because it is specific on `AuthScopeFlags`

arthur740212 replied:
Actually it can't because it needs to implement CanParse from its base

paulhazen replied:
I think that `ParsesToEnumFieldValidatorAttribute` can be removed altogether - there should not be a need for it since there are no longer any `enum` values that are represented by `List<string>`.

paulhazen commented on com.playeveryware.eos/Runtime/Core/Config/Attributes/ParsesToEnumFieldValidatorAttribute.cs:
```diff
@@ -0,0 +1,67 @@
+
+    [AttributeUsage(AttributeTargets.Field)]
+
+    public abstract class ParsesToEnumFieldValidatorAttribute : FieldValidatorAttribute
+    {
+        public const string FailedToParseTokensMessage = "Failed to parse all provided values into destination enum flags.";
+
+        public sealed override bool FieldValueIsValid(object toValidate, out string configurationProblemMessage)
+        {
+            List<string> flags = new List<string>();
+
+            if ((toValidate is string singleStringValue))
+            {
+                flags.Add(singleStringValue);
+            }
+            else if ((toValidate is List<string> listValues))
+            {
+                flags = new List<string>(listValues);
+            }
+            else
+            {
+                configurationProblemMessage = $"Value is neither a string or a List<string>.";
+                return false;
+            }
+
+            if (!CanParse(flags))
+            {
+                configurationProblemMessage = FailedToParseTokensMessage;
+                return false;
+            }
+
+            configurationProblemMessage = string.Empty;
+            return true;
+        }
+
+        protected abstract bool CanParse(List<string> flags);
+    }
```
As with the previous file - this one should no longer be needed.

arthur740212 replied:
This will be kept for now because it is a generic enum parse test. And we only dealt with EOSConfig enums, while other enums might not have the utility ready.

paulhazen commented on Assets/Tests/PlayMode/EOSConfigTests.cs:
I believe that this should be moved to `Assets/Tests/PlayMode/Config/EOSConfigTests.cs`

paulhazen commented on com.playeveryware.eos/Runtime/Core/Config/EOSConfig.cs:
```diff
@@ -200,6 +212,7 @@ protected EOSConfig() : base("EpicOnlineServicesConfig.json")
             3)]
         [JsonConverter(typeof(ListOfStringsToAuthScopeFlags))]
+        [ParsesToAuthScopeFieldValidator]
         public AuthScopeFlags authScopeOptionsFlags;
```
This should no longer need the field validator. None of the field members that are `enum` types should need them, since it is not possible to have an invalid value for them - unless the JSON itself is edited - and in that case the error should be generated from the JSON data layer.

arthur740212 replied:
This is kept and the test was changed to test for invalid input instead

paulhazen commented on com.playeveryware.eos/Runtime/Core/Config/Attributes/DevelopmentEnvironmentFieldValidatorAttribute.cs:
```diff
@@ -0,0 +1,88 @@
+                configurationProblemMessage = string.Empty;
+                return true;
+            }
+
+            if (Regex.IsMatch(singleStringValue, PreProductionEnvironmentRegex))
+            {
+                configurationProblemMessage = string.Empty;
+                return true;
+            }
+
+            if (Guid.TryParse(singleStringValue, out _))
+            {
+                configurationProblemMessage = string.Empty;
+                return true;
+            }
+
+            configurationProblemMessage = FieldDidNotMatchMessage;
+            return false;
+        }
+
+        bool isValidEntry(string toValidate)
+        {
+            if (string.IsNullOrEmpty(toValidate))
+            {
+                return true;
+            }
+
+            if (Regex.IsMatch(toValidate, PreProductionEnvironmentRegex))
+            {
+                return true;
+            }
+
+            if (Guid.TryParse(toValidate, out _))
+            {
+                return true;
+            }
+
+            return false;
+        }
+    }
```
I believe that this class should be named `SandboxFieldValidator` or something that makes it clearer what it's purpose is.

Additionally, what you could do to reduce the duplicated code between this class and the GUID validator, is you could have this class _extend_ the GUID validator, and the override could call the base implementation from the GUID validator, but then also test for the private pattern.

paulhazen commented on com.playeveryware.eos/Runtime/Core/Config/Attributes/DevelopmentEnvironmentFieldValidatorAttribute.cs:
```diff
@@ -0,0 +1,88 @@
+            return false;
+        }
+
+        bool isValidEntry(string toValidate)
```
Is this function utilized anywhere? It seems to do similar things to the function defined above it.

andrew-hirata-playeveryware replied:
If this isn't used, can we remove it?

arthur740212 committed:
fix(config,test) : changes on config validator from PR

Fix comment on ExpandFieldAttribute
Move EOSConfigTests to Config folder
Rename DevelopmentFieldValidatorAttribute to SandboxIDFieldValidatorAttribute
Remove success case if sandboxID is empty
Morph the enum parsing test into a test for invalid input values

paulhazen commented on com.playeveryware.eos/Runtime/Core/Config/Attributes/ParsesToEnumFieldValidatorAttribute.cs:
```diff
@@ -0,0 +1,68 @@
+
+    [AttributeUsage(AttributeTargets.Field)]
+
+    public abstract class ParsesToEnumFieldValidatorAttribute : FieldValidatorAttribute
+    {
+        public const string FailedToParseTokensMessage = "Failed to parse all provided values into destination enum flags.";
+        public const string ValueUnparsableMessage = "Value is neither a string or a List<string>.";
+
+        public sealed override bool FieldValueIsValid(object toValidate, out string configurationProblemMessage)
+        {
+            List<string> flags = new List<string>();
+
+            if ((toValidate is string singleStringValue))
+            {
+                flags.Add(singleStringValue);
+            }
+            else if ((toValidate is List<string> listValues))
+            {
+                flags = new List<string>(listValues);
+            }
+            else
+            {
+                configurationProblemMessage = ValueUnparsableMessage;
+                return false;
+            }
+
+            if (!CanParse(flags))
+            {
+                configurationProblemMessage = FailedToParseTokensMessage;
+                return false;
+            }
+
+            configurationProblemMessage = string.Empty;
+            return true;
+        }
+
+        protected abstract bool CanParse(List<string> flags);
+    }
```
I do not believe that this class is needed any longer.

paulhazen commented on com.playeveryware.eos/Runtime/Core/Config/Attributes/ParsesToPlatformFlagFieldValidatorAttribute.cs:
```diff
@@ -0,0 +1,47 @@
+
+    [AttributeUsage(AttributeTargets.Field)]
+
+    public class ParsesToPlatformFlagFieldValidatorAttribute : ParsesToEnumFieldValidatorAttribute
+    {
```
I do not believe that this class is needed any longer.

paulhazen commented on com.playeveryware.eos/Runtime/Core/Config/Attributes/ParsesToUlongFieldValidatorAttribute.cs:
```diff
@@ -0,0 +1,58 @@
+
+    [AttributeUsage(AttributeTargets.Field)]
+
+    public class ParsesToUlongFieldValidatorAttribute : FieldValidatorAttribute
```
I do not believe that this class is needed any longer.

paulhazen commented on com.playeveryware.eos/Runtime/Core/Config/Attributes/SandboxIDFieldValidatorAttribute.cs:
```diff
@@ -0,0 +1,68 @@
+                return false;
+            }
+
+            if (string.IsNullOrEmpty(singleStringValue))
+            {
+                configurationProblemMessage = string.Empty;
+                return true;
+            }
```
This implies to me that `SandboxID` _can_ be null or empty - but I do not think that this is the case.

WispyMouse replied:
Correct, that's what this logic does. Looking at the original spec of the task, that's not one of the things this should allow. So we'll make sure to change this soon to not allow empties.

arthur740212 replied:
Done

arthur740212 committed:
remove(config,test) : trim redundant parsing tests from config unit tests
